model/diffusion/diffusion_ibrl.py

```diff
@@ -0,0 +1,286 @@
+
+            # get the BC Q value
+            next_q1_bc = self.target_networks[q1_ind](next_obs, next_actions_bc)
+            next_q2_bc = self.target_networks[q2_ind](next_obs, next_actions_bc)
+            next_q_bc = torch.min(next_q1_bc, next_q2_bc)
+
+            # get the RL Q value
+            next_q1_rl = self.target_networks[q1_ind](next_obs, next_actions_rl)
+            next_q2_rl = self.target_networks[q2_ind](next_obs, next_actions_rl)
+            next_q_rl = torch.min(next_q1_rl, next_q2_rl)
+
+            # take the max Q value
+            next_q = torch.where(next_q_bc > next_q_rl, next_q_bc, next_q_rl)
+
+            # target value
+            target_q = rewards + gamma * (1 - terminated) * next_q  # (B,)
+
+        # run all critics in batch
+        current_q = torch.vmap(self.critic_wrapper, in_dims=(0, 0, None))(
+            self.ensemble_params, self.ensemble_buffers, (obs, actions)
+        )  # (n_critics, B)
+        loss_critic = torch.mean((current_q - target_q[None]) ** 2)
+        return loss_critic
+
+    def loss_actor(self, obs):
+        action = self.forward_train(
+            obs,
+            deterministic=False,
+        )  # use online policy only, also IBRL does not use tanh squashing
+        current_q = torch.vmap(self.critic_wrapper, in_dims=(0, 0, None))(
+            self.ensemble_params, self.ensemble_buffers, (obs, action)
+        )  # (n_critics, B)
+        current_q = current_q.min(
+            dim=0
+        ).values  # unlike RLPD, IBRL uses the min Q value for actor update
+        loss_actor = -torch.mean(current_q)
+        return loss_actor
+
+    def update_target_critic(self, tau):
+        """need to use ensemble_params instead of critic_networks"""
+        for target_ind, target_critic in enumerate(self.target_networks):
+            for target_param_name, target_param in target_critic.named_parameters():
+                source_param = self.ensemble_params[target_param_name][target_ind]
+                target_param.data.copy_(
+                    target_param.data * (1.0 - tau) + source_param.data * tau
+                )
+
+    def update_target_actor(self, tau):
+        for target_param, source_param in zip(
+            self.target_actor.parameters(), self.network.parameters()
+        ):
+            target_param.data.copy_(
+                target_param.data * (1.0 - tau) + source_param.data * tau
+            )
+
+    # ---------- Sampling ----------#
+
+    def forward(
+        self,
+        cond,
+        deterministic=False,
+        reparameterize=False,
+    ):
+        """use both pre-trained and online policies"""
+        q1_ind, q2_ind = self.get_random_indices()
+
+        # sample an action from the BC policy
+        bc_action = self.forward_sample(
+            cond=cond,
+            deterministic=True,
+            network_override=self.bc_policy,
+        )
+
+        # sample an action from the RL policy
+        rl_action = super().forward(
```

Comment:
probably use self.forward_sample() instead? no need to use the parent class function?

Reply:
Done.
<!-- Reviewable comment -OBhfDoG0RQ_j_dmG-oh:b-896fix -->
<!-- Sent from Reviewable.io -->